lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+    def release(self) -> Optional[str]:
+        """Get release of package.
+
+        Returns:
+            (Optional[str]):
+                Release of package. Defaults to None if package is absent.
+        """
+        return self.__data.get("release", None)
+
+    @property
+    def repo(self) -> Optional[str]:
+        """Get repository package is from.
+
+        Returns:
+            (Optional[str]):
+                Repository of package: Defaults to None if package is absent.
+        """
+        return self.__data.get("repo", None)
+
+
+def _load_cache(
+    cache_type: Literal["installed", "available"]
+) -> List[Dict[str, Union[str, _PackageState]]]:
+    """Load DNF cache information.
+
+    Args:
+        cache_type (Literal["installed", "available"]):
+            Cache information category to load.
+
+    Raises:
+        DNFExecutionError: Raised if incorrect cache type is passed.
+
+    Returns:
+        (List[Dict[str, Union[str, _PackageState]]]):
+            Loaded cache information.
+    """
+    if cache_type not in ["installed", "available"]:
+        raise DNFExecutionError(
+            f"Cache type must be either installed or available, not {cache_type}"
+        )
+
+    result = []
+    state = {"installed": _PackageState.INSTALLED, "available": _PackageState.AVAILABLE}
+    for pkg_name, pkg_version, pkg_repo in itertools.filterfalse(
+        lambda x: len(x) != 3,
+        [
+            line.strip("\n").strip().split()
+            for line in StringIO(
+                _dnf("list", postargs=[f"--{cache_type}"]).stdout.strip("\n"),
+            )
+        ],
+    ):
+        placeholder = pkg_name.split(".")
+        epoch = None  # Not every DNF package has an epoch.
+        charset, buffer = deque(itertools.chain(pkg_version)), []
+        while charset:
+            char = charset.popleft()
+            if char == ":":
+                epoch = "".join(buffer)
+                buffer.clear()
+            elif char == "-":
+                version = "".join(buffer)
+                buffer.clear()
+            else:
+                buffer.append(char)
+
+        result.append(
+            {
+                "name": ".".join(placeholder[:-1]),
+                "arch": ".".join(placeholder[-1:]),
+                "epoch": epoch,
+                "version": version,
+                "release": "".join(buffer),
+                "repo": pkg_repo[1:] if pkg_repo.startswith("@") else pkg_repo,
+                "state": state[cache_type],
+            }
+        )
+
+    return result
+
+
+class _Cache:
+    """Track the current state of the DNF package cache."""
+
+    def __new__(cls) -> "_Cache":
+        """Create new  _Cache object.
+
+        Returns:
+            (_Cache): New _Cache object.
+        """
+        if not hasattr(cls, f"{cls.__name__}__instance"):
+            cls.__instance = super(_Cache, cls).__new__(cls)
+            cls.__instance.__initialized = False
+        return cls.__instance
```

Comment:
I can use that approach since `_Cache` is not meant to be directly accessed by end-users. I mostly did it this way because I was like "I only want one of these to exist at a time; ah I'll use the Singleton pattern!"

Reply:
Yeah, that's understandable. Often those "design patterns" come from the Java world, where you sort of need to do it differently. But I think in Python most people just use the module as a singleton. See for example https://stackoverflow.com/a/31887/68707

Reply:
`_Cache` was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb